tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/process/traversal/step/sideEffect/TinkerGraphStep.java

```diff
@@ -132,8 +133,20 @@ private <E extends Element> Iterator<E> iteratorList(final Iterator<E> iterator)
         try {
             while (iterator.hasNext()) {
                 final E e = iterator.next();
-                if (HasContainer.testAll(e, this.hasContainers))
-                    list.add(e);
+                try {
+                    if (HasContainer.testAll(e, this.hasContainers))
+                        list.add(e);
+                } catch (GremlinTypeErrorException ex) {
+                    if (getTraversal().isRoot()) {
+                        /*
+                         * If we are at the top level of the query, we do a binary reduction from
+                         * ERROR -> FALSE and filter the solution quietly.
+                         */
+                    } else {
+                        // not a ternary -> binary reducer, pass the ERROR on
+                        throw ex;
```

Comment:
What happens if the GremlinTypeErrorException happens in a non-root traversal like inside a Repeat? There I think we should reduce. It's only when the parent traversal is a value expression (like Not) that we should not reduce.

Reply:
I agree with that. I will put together an update which also checks the parent step of non-root traversals and triggers a reduction if the parent is not a value expression.

Reply:
I've pushed some changes which modify the error propagation rules such that the error will propagate if and only if the parent step is a FilterStep. I haven't completely convinced myself that this is the right rule yet. I believe more tests are still needed.